src/main/java/io/cdap/plugin/googleads/source/batch/GoogleAdsRecordReader.java

```
@@ -0,0 +1,187 @@
+          .withClientCustomerId(googleAdsBatchSourceConfig.getClient_customer_id())
+          .withDeveloperToken(googleAdsBatchSourceConfig.getDeveloper_token())
+          .withOAuth2Credential(credential)
+          .build();
+    } catch (ValidationException e) {
+      e.printStackTrace();
+    }
+
+    AdWordsServicesInterface adWordsServices = AdWordsServices.getInstance();
+
+
+    Selector selector = new Selector();
+    selector.getFields().addAll(Arrays.asList(googleAdsBatchSourceConfig.getReport_fields()));
+
+
+    // Create report definition.
+    ReportDefinition reportDefinition = new ReportDefinition();
+    reportDefinition.setReportName("Criteria performance report #" + System.currentTimeMillis());
+    reportDefinition.setDateRangeType(ReportDefinitionDateRangeType.LAST_30_DAYS);//todo CUSTOM_DATE selection
+    reportDefinition.setReportType(ReportDefinitionReportType.fromValue(googleAdsBatchSourceConfig.getReport_type()));
+    reportDefinition.setDownloadFormat(DownloadFormat.CSV);
+
+    ReportingConfiguration reportingConfiguration =
+      new ReportingConfiguration.Builder()
+        .skipReportHeader(!googleAdsBatchSourceConfig.getIncludeReportHeader())
+        .skipColumnHeader(!googleAdsBatchSourceConfig.getIncludeColumnHeader())
+        .skipReportSummary(!googleAdsBatchSourceConfig.getIncludeReportSummary())
+        .useRawEnumValues(googleAdsBatchSourceConfig.getUseRawEnumValues())
+        .includeZeroImpressions(googleAdsBatchSourceConfig.getIncludeZeroImpressions())
+        .build();
+    session.setReportingConfiguration(reportingConfiguration);
+
+    reportDefinition.setSelector(selector);
+
+    ReportDownloaderInterface reportDownloader =
+      adWordsServices.getUtility(session, ReportDownloaderInterface.class);
+
+
+    try {
+      ReportDownloadResponse response = reportDownloader.downloadReport(reportDefinition);
+      report = response.getAsString();
+    } catch (ReportException e) {
+      e.printStackTrace();
```

don't ignore exceptions /  retry logic.